File: tests/TokenTest.php
<?php
/**
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * @author Vincent Tscherter <tscherter@karmin.ch>
 * @author Sven Strittmatter <ich@weltraumschaf.de>
 */

namespace de\weltraumschaf\ebnf;

require_once "Token.php";

/**
 * Testcase for class {Token}.
 */
class TokenTest extends \PHPUnit_Framework_TestCase {

    public function testGetTypeAsString() {
        $p = new Position(5, 10);
        $t = new Token(Token::OPERATOR, "", $p);
        $this->assertEquals("OPERATOR", $t->getTypeAsString());
        $t = new Token(Token::IDENTIFIER, "", $p);
        $this->assertEquals("IDENTIFIER", $t->getTypeAsString());
        $t = new Token(Token::COMMENT, "", $p);
        $this->assertEquals("COMMENT", $t->getTypeAsString());
        $t = new Token(Token::LITERAL, "", $p);
        $this->assertEquals("LITERAL", $t->getTypeAsString());
        $t = new Token(Token::EOF, "", $p);
        $this->assertEquals("EOF", $t->getTypeAsString());
        $t = new Token(4711, "", $p);
        $this->assertEquals("(4711)", $t->getTypeAsString());
    }

    public function testToString() {
        $p = new Position(5, 10);
        $t = new Token(Token::IDENTIFIER, "ident", $p);
        $this->assertEquals("<'ident', IDENTIFIER, (5, 10)>", $t->__toString());
        $p = new Position(5, 10, "/foo/bar.ebnf");
        $t = new Token(Token::IDENTIFIER, "ident", $p);
        $this->assertEquals("<'ident', IDENTIFIER, /foo/bar.ebnf (5, 10)>", $t->__toString());
        $p = new Position(5, 10, "/foo/bar.ebnf");
        $t = new Token(Token::IDENTIFIER, "", $p);
        $this->assertEquals("<IDENTIFIER, /foo/bar.ebnf (5, 10)>", $t->__toString());
    }

    public function testIsType() {
        $p = new Position(5, 10);
        $t = new Token(Token::OPERATOR, "", $p);
        $this->assertTrue($t->isType(Token::OPERATOR));
        $this->assertFalse($t->isType(Token::IDENTIFIER));
        $this->assertFalse($t->isType(Token::COMMENT));
        $this->assertFalse($t->isType(Token::LITERAL));
        $this->assertFalse($t->isType(Token::EOF));

        $t = new Token(Token::IDENTIFIER, "", $p);
        $this->assertTrue($t->isType(Token::IDENTIFIER));
        $this->assertFalse($t->isType(Token::OPERATOR));
        $this->assertFalse($t->isType(Token::COMMENT));
        $this->assertFalse($t->isType(Token::LITERAL));
        $this->assertFalse($t->isType(Token::EOF));

        $t = new Token(Token::COMMENT, "", $p);
        $this->assertTrue($t->isType(Token::COMMENT));
        $this->assertFalse($t->isType(Token::IDENTIFIER));
        $this->assertFalse($t->isType(Token::OPERATOR));
        $this->assertFalse($t->isType(Token::LITERAL));
        $this->assertFalse($t->isType(Token::EOF));

        $t = new Token(Token::LITERAL, "", $p);
        $this->assertTrue($t->isType(Token::LITERAL));
        $this->assertFalse($t->isType(Token::IDENTIFIER));
        $this->assertFalse($t->isType(Token::COMMENT));
        $this->assertFalse($t->isType(Token::OPERATOR));
        $this->assertFalse($t->isType(Token::EOF));

        $t = new Token(Token::EOF, "", $p);
        $this->assertTrue($t->isType(Token::EOF));
        $this->assertFalse($t->isType(Token::IDENTIFIER));
        $this->assertFalse($t->isType(Token::COMMENT));
        $this->assertFalse($t->isType(Token::LITERAL));
        $this->assertFalse($t->isType(Token::OPERATOR));

        $t = new Token(4711, "", $p);
        $this->assertFalse($t->isType(Token::OPERATOR));
        $this->assertFalse($t->isType(Token::IDENTIFIER));
        $this->assertFalse($t->isType(Token::COMMENT));
        $this->assertFalse($t->isType(Token::LITERAL));
        $this->assertFalse($t->isType(Token::EOF));
    }

    public function testIsEqual() {
        $p = new Position(5, 10);
        $t = new Token(Token::LITERAL, "foo", $p);
        $this->assertTrue($t->isEqual("foo"));
        $this->assertFalse($t->isEqual("bar"));

        $this->assertFalse($t->isNotEqual("foo"));
        $this->assertTrue($t->isNotEqual("bar"));
    }

    public function testIsEquals() {
        $tokens = array(
            new Token(0, "a", new Position(0, 0)),
            new Token(0, "b", new Position(0, 0)),
            new Token(0, "c", new Position(0, 0)),
        );

        foreach ($tokens as $token) {
            /* @var $token Token */
            $this->assertTrue($token->isEquals(array("a", "b", "c")), $token->getValue());
            $this->assertFalse($token->isEquals(array("x", "y", "z")), $token->getValue());
            $this->assertFalse($token->isNotEquals(array("a", "b", "c")), $token->getValue());
            $this->assertTrue($token->isNotEquals(array("x", "y", "z")), $token->getValue());
        }
    }

    public function testUnquoteString() {
        $this->assertEquals('a test string', Token::unquoteString('"a test string"'));
        $this->assertEquals('a "test" string', Token::unquoteString('"a "test" string"'));
        $this->assertEquals('a "test" string', Token::unquoteString('"a \"test\" string"'));
        $this->assertEquals("a test string", Token::unquoteString("'a test string'"));
        $this->assertEquals("a 'test' string", Token::unquoteString("'a 'test' string'"));
        $this->assertEquals("a 'test' string", Token::unquoteString("'a \'test\' string'"));

        $t = new Token(0, '"a \"test\" string"', new Position(0, 0));
        $this->assertEquals('"a \"test\" string"', $t->getValue());
        $this->assertEquals('a "test" string', $t->getValue(true));
    }

    public function testGetPostion() {
        $t = new Token(0, "abc", new Position(1, 5));
        $end = $t->getPosition(true);
        $this->assertInstanceOf("de\weltraumschaf\ebnf\Position", $end);
        $this->assertEquals(1, $end->getLine());
        $this->assertEquals(8, $end->getColumn());
    }
}

File: java/src/test/java/de/weltraumschaf/ebnf/TokenTest.java
package de.weltraumschaf.ebnf;

import java.util.Arrays;
import java.util.List;
import static org.junit.Assert.*;
import org.junit.Test;

/**
 * Unit test for Token.
 *
 * @author Sven Strittmatter <weltraumschaf@googlemail.com>
 */
public class TokenTest {

    @Test public void testGetTypeAsString() {
        final Position pos = new Position(5, 10);
        Token token;
        token = new Token(TokenType.IDENTIFIER, "", pos);
        assertEquals("IDENTIFIER", token.getType().toString());
        token = new Token(TokenType.COMMENT, "", pos);
        assertEquals("COMMENT", token.getType().toString());
        token = new Token(TokenType.LITERAL, "", pos);
        assertEquals("LITERAL", token.getType().toString());
        token = new Token(TokenType.EOF, "", pos);
        assertEquals("EOF", token.getType().toString());
        token = new Token(TokenType.L_BRACK, "", pos);
        assertEquals("L_BRACK", token.getType().toString());
    }

    @Test public void testToString() {
        Position pos = new Position(5, 10);
        Token token;
        token = new Token(TokenType.IDENTIFIER, "ident", pos);
        assertEquals("<'ident', IDENTIFIER, (5, 10)>", token.toString());

        pos = new Position(5, 10, "/foo/bar.ebnf");
        token = new Token(TokenType.IDENTIFIER, "ident", pos);
        assertEquals("<'ident', IDENTIFIER, /foo/bar.ebnf (5, 10)>", token.toString());

        pos = new Position(5, 10, "/foo/bar.ebnf");
        token = new Token(TokenType.IDENTIFIER, "", pos);
        assertEquals("<IDENTIFIER, /foo/bar.ebnf (5, 10)>", token.toString());

        pos = new Position(5, 10, "/foo/bar.ebnf");
        token = new Token(TokenType.IDENTIFIER, "a-very-very-very-very-long-identifier", pos);
        assertEquals("<'a-very-very-ver...', IDENTIFIER, /foo/bar.ebnf (5, 10)>", token.toString());
    }

    @Test public void testIsType() {
        final Position pos = new Position(5, 10);
        Token token;

        token = new Token(TokenType.L_BRACK, "", pos);
        assertTrue(token.isType(TokenType.L_BRACK));
        assertFalse(token.isType(TokenType.IDENTIFIER));
        assertFalse(token.isType(TokenType.COMMENT));
        assertFalse(token.isType(TokenType.LITERAL));
        assertFalse(token.isType(TokenType.EOF));

        token = new Token(TokenType.IDENTIFIER, "", pos);
        assertTrue(token.isType(TokenType.IDENTIFIER));
        assertFalse(token.isType(TokenType.L_BRACK));
        assertFalse(token.isType(TokenType.COMMENT));
        assertFalse(token.isType(TokenType.LITERAL));
        assertFalse(token.isType(TokenType.EOF));

        token = new Token(TokenType.COMMENT, "", pos);
        assertTrue(token.isType(TokenType.COMMENT));
        assertFalse(token.isType(TokenType.IDENTIFIER));
        assertFalse(token.isType(TokenType.L_BRACK));
        assertFalse(token.isType(TokenType.LITERAL));
        assertFalse(token.isType(TokenType.EOF));

        token = new Token(TokenType.LITERAL, "", pos);
        assertTrue(token.isType(TokenType.LITERAL));
        assertFalse(token.isType(TokenType.IDENTIFIER));
        assertFalse(token.isType(TokenType.COMMENT));
        assertFalse(token.isType(TokenType.L_BRACK));
        assertFalse(token.isType(TokenType.EOF));

        token = new Token(TokenType.EOF, "", pos);
        assertTrue(token.isType(TokenType.EOF));
        assertFalse(token.isType(TokenType.IDENTIFIER));
        assertFalse(token.isType(TokenType.COMMENT));
        assertFalse(token.isType(TokenType.LITERAL));
        assertFalse(token.isType(TokenType.L_BRACK));
    }

    @Test public void testIsEqual() {
        final Position pos = new Position(5, 10);
        final Token token = new Token(TokenType.LITERAL, "foo", pos);
        assertTrue(token.isEqual("foo"));
        assertFalse(token.isEqual("bar"));

        assertFalse(token.isNotEqual("foo"));
        assertTrue(token.isNotEqual("bar"));
    }

    @Test public void testIsEquals() {
        final List<Token> tokens = Arrays.asList(
            new Token(TokenType.IDENTIFIER, "a", new Position(0, 0)),
            new Token(TokenType.IDENTIFIER, "b", new Position(0, 0)),
            new Token(TokenType.IDENTIFIER, "c", new Position(0, 0))
        );

        final String[] set1 = {"a", "b", "c"};
        final String[] set2 = {"x", "y", "z"};
        for (Token token : tokens) {
            assertTrue(token.isEquals(set1));
            assertFalse(token.isEquals(set2));
            assertFalse(token.isNotEquals(set1));
            assertTrue(token.isNotEquals(set2));
        }
    }

   @Test  public void testUnquoteString() {
        assertEquals("a test string", Token.unquoteString("\"a test string\""));
        assertEquals("a \"test\" string", Token.unquoteString("\"a \"test\" string\"")); // NOPMD
        assertEquals("a \"test\" string", Token.unquoteString("\"a \"test\" string\""));
        assertEquals("a test string", Token.unquoteString("'a test string'"));
        assertEquals("a 'test' string", Token.unquoteString("'a 'test' string'"));
        assertEquals("a 'test' string", Token.unquoteString("'a \'test\' string'"));

        final Token token = new Token(TokenType.COMMENT, "\"a \"test\" string\"", new Position(0, 0));
        assertEquals("\"a \"test\" string\"", token.getValue());
        assertEquals("a \"test\" string", token.getValue(true));
    }

    @Test public void testGetPostion() {
        final Token token = new Token(TokenType.IDENTIFIER, "abc", new Position(1, 5));
        final Position end = token.getPosition(true);
        assertEquals(1, end.getLine());
        assertEquals(8, end.getColumn());
    }

    @Test public void testIsOperator() {
        List<TokenType> types;

        types = Arrays.asList(
            TokenType.ASIGN, TokenType.CHOICE, TokenType.END_OF_RULE, TokenType.L_BRACE, TokenType.L_BRACK,
            TokenType.L_PAREN, TokenType.RANGE, TokenType.R_BRACE, TokenType.R_BRACK, TokenType.R_PAREN
        );

        final Position pos = new Position(0, 0);
        for (TokenType type : types) {
            final Token token = new Token(type, "", pos); // NOPMD
            assertTrue(token.isOperator());
        }

        types = Arrays.asList(
            TokenType.COMMENT, TokenType.EOF, TokenType.IDENTIFIER, TokenType.LITERAL
        );

        for (TokenType type : types) {
            Token token = new Token(type, "", pos); // NOPMD
            assertFalse(token.isOperator());
        }
    }
}
